driver-core/src/main/com/mongodb/internal/client/model/AbstractConstructibleBsonElementWrappingBson.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import static com.mongodb.internal.client.model.AbstractConstructibleBson.newMerged;
+import static java.lang.String.format;
+
+/**
+ * A {@link Bson} that contains exactly one name/value pair
+ * and allows constructing new instances via {@link #newWithAppendedValue(String, Object)} instead of mutating {@code this}.
+ * The value must itself be a {@code Bson}.
+ * <p>
+ * This class is conceptually the same as {@link AbstractConstructibleBsonElement},
+ * but unlike {@link AbstractConstructibleBsonElement}, it can be instantiated from another {@link Bson}
+ * that contains exactly one name/value pair.</p>
+ *
+ * @param <S> A type introduced by the concrete class that extends this abstract class.
+ * @see AbstractConstructibleBson
+ * @see AbstractConstructibleBsonElement
+ */
+public abstract class AbstractConstructibleBsonElementWrappingBson<S extends AbstractConstructibleBsonElementWrappingBson<S>>
+        implements Bson {
+    private static final Document EMPTY_APPENDED_ELEMENT_VALUE = new Document();
+
+    private final Bson baseElement;
+    private final AbstractConstructibleBson<?> appendedElementValue;
+
+    protected AbstractConstructibleBsonElementWrappingBson(final Bson baseElement) {
+        this(baseElement, EMPTY_APPENDED_ELEMENT_VALUE);
+    }
+
+    protected AbstractConstructibleBsonElementWrappingBson(final Bson baseElement, final Bson appendedElementValue) {
+        this.baseElement = baseElement;
+        this.appendedElementValue = AbstractConstructibleBson.of(appendedElementValue);
+    }
+
+    protected abstract S newSelf(Bson baseElement, Bson appendedElementValue);
+
+    /**
+     * {@linkplain Document#append(String, Object) Appends} the specified mapping to the value via {@link #newWithMutatedValue(Consumer)}.
+     *
+     * @return A new instance.
+     */
+    protected final S newWithAppendedValue(final String name, final Object value) {
+        return newWithMutatedValue(doc -> doc.append(name, value));
+    }
+
+    /**
+     * Creates a copy of {@code this} with a value that is
+     * a {@linkplain AbstractConstructibleBson#newMutated(Consumer) shallow copy} of this value mutated via the specified {@code mutator}.
+     *
+     * @return A new instance.
+     * @see AbstractConstructibleBson#newMutated(Consumer)
+     */
+    protected final S newWithMutatedValue(final Consumer<Document> mutator) {
+        return newSelf(baseElement, appendedElementValue.newMutated(mutator));
+    }
+
+    @Override
+    public final <TDocument> BsonDocument toBsonDocument(final Class<TDocument> documentClass, final CodecRegistry codecRegistry) {
+        BsonDocument baseElementDoc = baseElement.toBsonDocument(documentClass, codecRegistry);
+        if (baseElementDoc.size() != 1) {
+            throw new IllegalStateException(format("baseElement must contain exactly one element, but contains %s", baseElementDoc.size()));
```

Comment:
`baseElement` may come from user code, like it does, for example, in `SearchOperator.of`. While that is unquestionably a bug in the user code, for those we throw subclasses of `RuntimeException` instead of `AssertionError`s. That's why I chose `IllegalStateException`.